# 01 Data Cleaning Starter

Project: **AI Support Operations SLA Breach and Priority Triage**

## Objective
Inspect the raw support ticket dataset and create a cleaned sample suitable for EDA and modelling. Document every decision.

In [4]:
from pathlib import Path

print("Current Working Directory:")
print(Path.cwd())

Current Working Directory:
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\02_Notebooks


In [5]:
from pathlib import Path

for item in Path.cwd().iterdir():
    print(item)

C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\02_Notebooks\.ipynb_checkpoints
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\02_Notebooks\01_data_cleaning_starter.ipynb
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\02_Notebooks\02_eda_starter.ipynb
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\02_Notebooks\03_feature_engineering_starter.ipynb
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\02_Notebooks\04_model_training_starter.ipynb
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\02_Notebooks\05_evaluation_pipeline_starter.ipynb
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\02_Notebooks\06_final_ai_engineer_report_starter.ipynb


In [6]:
from pathlib import Path

print(Path.cwd().parent)

for item in Path.cwd().parent.iterdir():
    print(item)

C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\01_Data
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\02_Notebooks
C:\Users\jumma\Downloads\15_Day_AI_Engineer_Workplace_Assignment\03_Reports_and_Templates


In [7]:
from pathlib import Path

for item in Path("../01_Data").iterdir():
    print(item)

..\01_Data\processed data
..\01_Data\raw data


In [8]:
from pathlib import Path

for item in Path("../01_Data/raw data").iterdir():
    print(item)

..\01_Data\raw data\.ipynb_checkpoints
..\01_Data\raw data\data_dictionary.csv
..\01_Data\raw data\generate_large_dataset.py
..\01_Data\raw data\sample_dataset.csv


In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
DATA_PATH = Path("../01_Data/raw data/sample_dataset.csv")

df = pd.read_csv(DATA_PATH)
df.head()

,ticket_id,account_id,customer_segment,uk_region,support_channel,product_area,issue_category,priority_initial,submitted_at,day_of_week,...,assigned_agent,agent_experience_months,backlog_age_hours,first_response_minutes,reopened_last_90d,resolved_at,escalated_prior_to_resolution,resolution_hours,final_priority,sla_breached
0,TCK-202500000,ACC-93810,SMB,Wales,Email,Reporting,service_outage,Critical,2025-09-06 18:51:00,Saturday,...,Amelia,6,13.11,62,0,2025-09-07 20:56:43,1,26.10,Critical,1
1,TCK-202500001,ACC-82357,SMB,East Midlands,Web,Billing,data_quality,Medium,2026-02-27 08:29:00,Friday,...,Isla,30,8.35,289,0,2026-02-28 05:55:43,0,21.45,Medium,0
2,TCK-202500002,ACC-45093,SMB,East Midlands,Web,API Integration,access_request,Medium,2025-10-19 05:16:00,Sunday,...,Muhammad,19,11.28,215,0,2025-10-19 17:38:32,0,12.38,Medium,0
3,TCK-202500003,ACC-44973,Enterprise,Northern Ireland,Web,Workflow Automation,how_to,Critical,2025-12-01 03:05:00,Monday,...,Isla,25,20.45,261,0,2025-12-01 10:42:39,1,7.63,Critical,1
4,TCK-202500004,ACC-26483,SMB,South West,Email,Security,performance,High,2026-01-16 11:48:00,Friday,...,Harry,35,0.06,139,0,2026-01-16 23:11:33,0,11.39,High,0


## Initial inspection
Check shape, columns, data types, missing values, duplicates and target balance.

In [13]:
print("Shape:", df.shape)
display(df.info())
display(df.isna().mean().sort_values(ascending=False).head(15))
print("Duplicate rows:", df.duplicated().sum())
display(df["sla_breached"].value_counts(normalize=True))


Shape: (15090, 30)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15090 entries, 0 to 15089
Data columns (total 30 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   ticket_id                      15090 non-null  object 
 1   account_id                     15090 non-null  object 
 2   customer_segment               15090 non-null  object 
 3   uk_region                      15090 non-null  object 
 4   support_channel                15090 non-null  object 
 5   product_area                   14892 non-null  object 
 6   issue_category                 15090 non-null  object 
 7   priority_initial               14911 non-null  object 
 8   submitted_at                   15001 non-null  object 
 9   day_of_week                    15090 non-null  object 
 10  hour_of_day                    15090 non-null  int64  
 11  customer_tenure_months         14818 non-null  float64
 12  contract_value_gbp         

None

avg_sentiment_score              0.025447
customer_tenure_months           0.018025
product_area                     0.013121
priority_initial                 0.011862
submitted_at                     0.005898
first_response_minutes           0.000000
assigned_agent                   0.000000
agent_experience_months          0.000000
backlog_age_hours                0.000000
ticket_id                        0.000000
reopened_last_90d                0.000000
contains_refund_keyword          0.000000
resolved_at                      0.000000
escalated_prior_to_resolution    0.000000
resolution_hours                 0.000000
dtype: float64

Duplicate rows: 90


sla_breached
1    0.609278
0    0.390722
Name: proportion, dtype: float64

## Initial Inspection Summary

The dataset contains 15,090 support ticket records and 30 variables related to customer accounts, ticket characteristics, operational metrics, and SLA outcomes. Initial inspection was performed to understand the overall structure, data quality, and suitability of the dataset for predictive modelling.

Several columns contain missing values, with the highest proportion observed in `avg_sentiment_score` (2.54%), followed by `customer_tenure_months` (1.80%), `product_area` (1.31%), `priority_initial` (1.19%), and `submitted_at` (0.59%). The relatively low proportion of missing data suggests that imputation and targeted cleaning strategies can be applied without significant information loss.

A total of 90 duplicate records were identified, indicating that duplicate removal will be required during the data-cleaning stage to avoid bias in subsequent analysis and model training.

The target variable, `sla_breached`, shows a distribution of approximately 60.9% breached tickets and 39.1% non-breached tickets. This indicates a moderately balanced classification problem and suggests that standard evaluation metrics such as Precision, Recall, F1-Score, and ROC-AUC can be used without immediately requiring advanced resampling techniques.

Overall, the dataset appears suitable for machine learning after addressing identified data-quality issues, including missing values, duplicate records, inconsistent categorical labels, and potential data leakage variables.


## Data-quality checks
Look for inconsistent categories, invalid numerical values and leakage-risk columns.

In [15]:
# ============================================================
# Data Quality Checks
# ============================================================

categorical_columns = [
    "uk_region",
    "support_channel",
    "product_area",
    "priority_initial"
]

for col in categorical_columns:
    print(f"\n{'='*50}")
    print(f"Column: {col}")
    print(f"{'='*50}")

    display(
        df[col]
        .value_counts(dropna=False)
        .head(20)
    )

# Check invalid queue values
negative_queue_count = (
    df["agent_queue_length_at_submit"] < 0
).sum()

print("\nNegative queue values:", negative_queue_count)

# Identify leakage-risk columns
leakage_columns = [
    "resolved_at",
    "escalated_prior_to_resolution",
    "resolution_hours",
    "final_priority"
]

print("\nLeakage-risk columns:")
for col in leakage_columns:
    print(f" - {col}")


Column: uk_region


uk_region
London                1538
South West            1485
Yorkshire             1481
Wales                 1456
West Midlands         1454
South East            1434
Scotland              1424
Greater Manchester    1418
Northern Ireland      1391
East Midlands         1380
east midlands           38
wales                   38
south west              37
west midlands           36
northern ireland        36
south west              35
south east              35
london                  32
london                  31
greater manchester      30
Name: count, dtype: int64


Column: support_channel


support_channel
Email             4331
Web               3980
Chat              2783
Phone             2450
Partner Portal     859
WEB                150
E-mail             140
email              134
chat               132
Partner portal     131
Name: count, dtype: int64


Column: product_area


product_area
Billing                2036
Workflow Automation    1954
Account Access         1921
Payments               1703
Reporting              1640
Mobile App             1461
API Integration        1422
Data Export            1183
Security               1152
NaN                     198
account access           71
reporting                51
workflow automation      49
api integration          47
billing                  47
mobile app               46
payments                 44
data export              39
security                 26
Name: count, dtype: int64


Column: priority_initial


priority_initial
Critical    6831
High        4664
Medium      2713
Low          703
NaN          179
Name: count, dtype: int64


Negative queue values: 64

Leakage-risk columns:
 - resolved_at
 - escalated_prior_to_resolution
 - resolution_hours
 - final_priority


## Data Quality Assessment Summary

The data-quality review identified several issues that require remediation before modelling. Multiple categorical variables contain inconsistent labels caused by differences in capitalization and formatting. For example, the `support_channel` field contains values such as "Email", "email", and "E-mail", while the `uk_region` and `product_area` fields contain similar inconsistencies. These variations represent the same underlying categories and should be standardized to improve data quality and reduce unnecessary feature fragmentation.

The analysis also identified 64 records with negative values in `agent_queue_length_at_submit`. Since queue length cannot be negative in a real support environment, these values are considered invalid and will be corrected during cleaning.

Missing values were detected in several columns, including `avg_sentiment_score`, `customer_tenure_months`, `product_area`, `priority_initial`, and `submitted_at`. Although the percentage of missing values is relatively low, appropriate treatment is required to ensure data completeness and consistency.

Finally, four variables were identified as potential sources of target leakage: `resolved_at`, `escalated_prior_to_resolution`, `resolution_hours`, and `final_priority`. These features contain information that becomes available during or after ticket resolution and therefore must be excluded from predictive modelling to maintain realistic model performance and avoid leakage.


## Cleaning Decisions

The dataset was cleaned by standardizing categorical values, removing duplicate records, correcting invalid queue length values, converting date columns to datetime format, and handling missing values using simple and reproducible imputation methods. These decisions were made to improve data quality while minimizing information loss.

In [17]:
# ============================================================
# Data Cleaning
# ============================================================

clean = df.copy()

# ------------------------------------------------------------
# Standardise text categories
# ------------------------------------------------------------

for col in [
    "uk_region",
    "support_channel",
    "product_area",
    "priority_initial",
    "issue_category",
    "customer_segment"
]:
    clean[col] = clean[col].astype("string").str.strip()

# Standardise support channels
clean["support_channel"] = (
    clean["support_channel"]
    .str.lower()
    .replace({
        "e-mail": "email",
        "email": "email",
        "web": "web",
        "chat": "chat",
        "phone": "phone",
        "partner portal": "partner portal"
    })
)

# Standardise regions and product names
clean["uk_region"] = clean["uk_region"].str.title()
clean["product_area"] = clean["product_area"].str.title()

# ------------------------------------------------------------
# Convert dates
# ------------------------------------------------------------

clean["submitted_at"] = pd.to_datetime(
    clean["submitted_at"],
    errors="coerce"
)

clean["resolved_at"] = pd.to_datetime(
    clean["resolved_at"],
    errors="coerce"
)

# ------------------------------------------------------------
# Fix invalid values
# ------------------------------------------------------------

clean.loc[
    clean["agent_queue_length_at_submit"] < 0,
    "agent_queue_length_at_submit"
] = np.nan

# ------------------------------------------------------------
# Remove duplicate records
# ------------------------------------------------------------

clean = clean.drop_duplicates()

# ------------------------------------------------------------
# Handle missing numerical values
# ------------------------------------------------------------

clean["avg_sentiment_score"] = (
    clean["avg_sentiment_score"]
    .fillna(clean["avg_sentiment_score"].median())
)

clean["customer_tenure_months"] = (
    clean["customer_tenure_months"]
    .fillna(clean["customer_tenure_months"].median())
)

clean["agent_queue_length_at_submit"] = (
    clean["agent_queue_length_at_submit"]
    .fillna(clean["agent_queue_length_at_submit"].median())
)

# ------------------------------------------------------------
# Handle missing categorical values
# ------------------------------------------------------------

clean["product_area"] = (
    clean["product_area"]
    .fillna(clean["product_area"].mode()[0])
)

clean["priority_initial"] = (
    clean["priority_initial"]
    .fillna(clean["priority_initial"].mode()[0])
)

# ------------------------------------------------------------
# Validation Checks
# ------------------------------------------------------------

print("Shape after cleaning:", clean.shape)

print("\nRemaining Missing Values:")
display(
    clean.isna()
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

print("\nDuplicate Rows Remaining:")
print(clean.duplicated().sum())

print("\nNegative Queue Values Remaining:")
print(
    (clean["agent_queue_length_at_submit"] < 0).sum()
)

Shape after cleaning: (15000, 30)

Remaining Missing Values:


submitted_at                     0.005933
ticket_id                        0.000000
message_length                   0.000000
final_priority                   0.000000
resolution_hours                 0.000000
escalated_prior_to_resolution    0.000000
resolved_at                      0.000000
reopened_last_90d                0.000000
first_response_minutes           0.000000
backlog_age_hours                0.000000
dtype: float64


Duplicate Rows Remaining:
0

Negative Queue Values Remaining:
0


## Cleaning Results

The data-cleaning process successfully addressed the key quality issues identified during the initial assessment. Duplicate records were removed, reducing the dataset size from 15,090 to 15,000 observations. Inconsistent categorical labels were standardized to improve consistency across region, support channel, and product area variables.

Invalid queue length values were corrected and no negative queue values remained after cleaning. Missing numerical values were imputed using median values, while missing categorical values were replaced using the most frequently occurring category.

A small number of records (less than 1% of the dataset) contained missing submission timestamps. Because submission time is important for later feature engineering activities, these records will be removed to ensure data completeness.

Following cleaning, the dataset is considered suitable for exploratory analysis, feature engineering, and predictive modelling.


## Save cleaned sample

In [20]:
from pathlib import Path

OUTPUT_PATH = Path("../01_Data/processed data/cleaned_support_sla_sample.csv")

clean.to_csv(OUTPUT_PATH, index=False)

print("Saved:", OUTPUT_PATH)

Saved: ..\01_Data\processed data\cleaned_support_sla_sample.csv
